### PR TITLE
Enabling CMake OMNeT++ on Windows 

### PR DESCRIPTION
**Pull request by HpLightcorner**
This PR enables OMNeT++ CMake to be used on Windows. Also, some changes are applied to allow to work with OMNeT++ Version 6 (changing e.g. how messages are generated by copying the message description before generating code). 

A CMake feature used to correctly work with Windows paths bumps the minimum CMake version to 3.20 - this may need further discussion.

### CODE REVIEW

File: FindOmnetPP.cmake [hunk 1]
```diff
@@ -1,6 +1,7 @@
 find_path(OMNETPP_ROOT NAMES bin/omnetpp PATHS ENV PATH PATH_SUFFIXES .. DOC "Path to OMNeT++ root directory")
 find_path(OMNETPP_INCLUDE_DIR NAMES omnetpp.h PATHS ${OMNETPP_ROOT}/include DOC "OMNeT++ include directory")
-find_program(OMNETPP_MSGC NAMES nedtool opp_msgc PATHS ${OMNETPP_ROOT}/bin DOC "OMNeT++ message compiler")
+find_program(OMNETPP_MSGC NAMES opp_msgtool PATHS ${OMNETPP_ROOT}/bin DOC "OMNeT++ message compiler")
+find_program(OMNETPP_FEATURETOOL NAMES opp_featuretool PATHS ${OMNETPP_ROOT}/bin DOC "OMNeT++ Featuretool")
```
Comment (riebl) [inline FindOmnetPP.cmake]: Spelling: OMNeT++ feature tool?

File: AddOppRun.cmake [hunk 2]
```diff
@@ -61,12 +63,15 @@ function(_build_opp_run_command)
...
     get_target_property(target_type ${args_TARGET} TYPE)
     if(${target_type} STREQUAL "EXECUTABLE")
-        set(exec $<TARGET_FILE:${args_TARGET}> -n ${ned_folders})
+        set(exec $<TARGET_FILE:${args_TARGET}> -n "${ned_folders}")
     elseif(${target_type} STREQUAL "SHARED_LIBRARY")
-        set(exec ${opp_run} -n ${ned_folders} -l $<TARGET_FILE:${args_TARGET}>)
+        # if the target is depending on libraries like INET, 
+        # add them before the actual target so that they get loaded first by opp_run
+        _get_opp_run_libraries(${args_TARGET} opp_libraries)
+        set(exec ${opp_run} -n "${ned_folders}" ${opp_libraries} -l $<TARGET_FILE:${args_TARGET}>)
```
Comment (riebl) [inline AddOppRun.cmake:71]: I am fine with this change in general; however, it seems unrelated to the switch from colons to semi-colons.

File: GenerateOppMessage.cmake [hunk 3]
```diff
@@ -1,31 +1,99 @@
...
+# One Value Arguments:
+# - TARGET: Optional name of the target to call target_sources and target_include_directories to add compiled sources
+# - OUTPUT_ROOT: Optional root-directory where the files will be generated. Default: ${PROJECT_BINARY_DIR}/opp_messages
+# - DIRECTORY: Optional relative path for the output-directory (Appended to OUTPUT_ROOT). 
+#   Defaults to the relative path of the message file in respect to ${PROJECT_SOURCE_DIR}/src
+# - SRCS: Optional name of the variable to populate with sources to be compiled
+# - INCLUDE_DIR: Optional name of the variable to populate with the include directory
+#
+# Mutli Value Arguments:
+# - ADDITIONAL_NED_PATHS: Optional paths to be added as search/include directories when calling the message compiler (-I Arguments)
```
Comment (riebl) [inline GenerateOppMessage.cmake:15]: Maybe just "NED_PATHS"? Adding NED paths explicitly is always "additional", right?

File: GenerateOppMessage.cmake [hunk 4]
```diff
@@ -1,31 +1,99 @@
...
+
+    foreach(include_dir IN LISTS args_ADDITIONAL_NED_PATHS)
+        list(APPEND _args "-I" ${include_dir})
+    endforeach()
+
+    # handle Message-Version
+    if(args_MSG4)
+        list(APPEND _args "--msg4")
+    else()
+        list(APPEND _args "--msg6")
```
Comment (riebl) [inline GenerateOppMessage.cmake]: I would expect not to pass any *--msgX* option by default.

File: GenerateOppMessage.cmake [hunk 5]
```diff
@@ -1,31 +1,99 @@
...
 
+# Options:
+# - MSG4: Call message compiler with --msg4
+# 
+# One Value Arguments:
+# - TARGET: Optional name of the target to call target_sources and target_include_directories to add compiled sources
+# - OUTPUT_ROOT: Optional root-directory where the files will be generated. Default: ${PROJECT_BINARY_DIR}/opp_messages
+# - DIRECTORY: Optional relative path for the output-directory (Appended to OUTPUT_ROOT). 
+#   Defaults to the relative path of the message file in respect to ${PROJECT_SOURCE_DIR}/src
+# - SRCS: Optional name of the variable to populate with sources to be compiled
```
Comment (riebl) [inline GenerateOppMessage.cmake]: What is the more common name in CMake, *SRCS* or *SOURCES*? I know that *FindProtobuf.cmake* names its (positional) parameter *SRCS*. I just want to keep it consistent with other CMake code.

File: AddOppTarget.cmake [hunk 6]
```diff
@@ -52,15 +53,17 @@ function(add_opp_target)
...
-            WORKING_DIRECTORY ${msg_out_dir} VERBATIM)
 
-        list(APPEND cpp_files "${msg_out_dir}/${msg_name}_m.cc" "${msg_out_dir}/${msg_name}_m.h")
+        generate_opp_message(
+            ${msg_file}
+            OUTPUT_ROOT             ${msg_gen_dir}
+            DIRECTORY               ${msg_prefix}
+            SRCS                    _cpp_files
+            ADDITIONAL_NED_PATHS    ${args_SOURCE_DIR}
+            MSG4
```
Comment (riebl) [inline AddOppTarget.cmake]: You define *MSG4* as an option to `add_opp_target` but always set it here no matter what.

File: OMNet.cmake [hunk 7]
```diff
@@ -1,3 +1,5 @@
+cmake_minimum_required(VERSION 3.20)
```
Comment (riebl) [inline OMNet.cmake]: I am not happy with requiring 3.20 in general. With respect to Artery, my goal is to support the current Debian stable. At the moment Debian stable ships with 3.13.
Can we please make this a Windows-only requirement?

File: GetNedFolders.cmake [hunk 8]
```diff
@@ -1,4 +1,4 @@
-function(_get_ned_folders _target _property _base_dir _output)
+function(__get_ned_folders _target _property _base_dir _output)
```
Comment (riebl) [inline GetNedFolders.cmake]: Never had this bug before. Is this a well-known issue?

**Review by riebl — COMMENTED** [review]
I haven't tested yet if your changes still work for my Linux setup. So that is my to-do before approval.

File: AddOppRun.cmake (hunk 2 above)
Comment (HpLightcorner) [reply AddOppRun.cmake:71]: it is needed so that `opp_run` finds the libraries to load, at least on windows. Maybe not clear in the commit message, but necessary...

File: GenerateOppMessage.cmake (hunk 3 above)
Comment (HpLightcorner) [reply GenerateOppMessage.cmake:15]: To be honest: wanted to be explicit. CMake is already confusing enough. Of course, we can change, I somehow think it makes the CMake code slightly more readable.

File: GenerateOppMessage.cmake (hunk 4 above)
Comment (HpLightcorner) [reply GenerateOppMessage.cmake]: is the default `--msg6` already? Will check, but agree.

File: GenerateOppMessage.cmake (hunk 5 above)
Comment (HpLightcorner) [reply GenerateOppMessage.cmake]: Have seen SRCS a lot. @thor - your thoughts?

File: AddOppTarget.cmake (hunk 6 above)
Comment (HpLightcorner) [reply AddOppTarget.cmake]: yep, will fix that

File: GetNedFolders.cmake (hunk 8 above)
Comment (HpLightcorner) [reply GetNedFolders.cmake]: We already had a discussion on that topic, I am not sure why this is happening on my machine, but it IS happening. So, maybe we will have to file a bug report at CMake, this workaround at least solved my issues.

File: OMNet.cmake (hunk 7 above)
Comment (HpLightcorner) [reply OMNet.cmake]: Yes... I get your point, modern CMake gets more and more really useful features...
I guess there is some ugly workaround for the feature we need, let me investigate...

File: FindOmnetPP.cmake (hunk 1 above)
Comment (HpLightcorner) [reply FindOmnetPP.cmake]: Sorry, don't get your point? You mean `OMNETPP_FEATURE_TOOL`?

File: GenerateOppMessage.cmake (hunk 4 above)
Comment (HpLightcorner) [reply GenerateOppMessage.cmake]: yes, it defaults to `--msg6` according to `opp_msgtool` source code. Will remove the explicit `--msg6`.

File: GetNedFolders.cmake (hunk 8 above)
Comment (riebl) [reply GetNedFolders.cmake]: That is very strange because also official CMake modules such as *FindBoost* use a single underscore as the prefix for internal function names. Do you have any hints on how to reproduce this problem? Does this happen only on Windows machines? Any specific CMake version?

File: FindOmnetPP.cmake (hunk 1 above)
Comment (riebl) [reply FindOmnetPP.cmake]: No, I am referring to the DOC string. "feature tool" would be in line with "message compiler" etc.

**Comment by riebl** [comment]
I had to add `WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}` to the copy step in `generate_opp_message` to not receive a copying error.
Edit: Hold on, there seems to be more broken right now…

File: GetNedFolders.cmake (hunk 8 above)
Comment (HpLightcorner) [reply GetNedFolders.cmake]: CMake Version 3.20 - it is at least happening with my environment, Windows, VSCode etc. - I suggest we implement the workaround and I will open a report at CMake, maybe there is more behind it.

**Comment by HpLightcorner** [comment]
I for sure highly focused on version 6.x - so might be that I have overseen some features needed for the 5.x builds with the message generator.

File: GetNedFolders.cmake (hunk 8 above)
Comment (HpLightcorner) [reply GetNedFolders.cmake]: I had a look at FindBoost and I think I found a difference between FindBoost and our use case: There is no private function with the same public name like we use:
- Public: `my_func`
- Private: `_my_func`

So what seems to happen: CMake ignores the `_` and instead calls my_func for calls to `_my_func` causing a recursion error.

File: FindOmnetPP.cmake (hunk 1 above)
Comment (HpLightcorner) [reply FindOmnetPP.cmake]: Agree, will fix.

**Comment by riebl** [comment]
Seems to run fine again with my fixup commit. Vice versa, I hope it does not break your 6.x setup.
Adding `-I ${msg_dir}` seems to be unnecessary and even breaks if `${msg_dir}` is empty.

File: GetNedFolders.cmake (hunk 8 above)
Comment (HpLightcorner) [reply GetNedFolders.cmake]: I came up with the following test, basically reflecting what we are doing:
```
cmake_minimum_required(VERSION 3.20)

project(top)

function(_my_func CALL_MYSELF)
    message(STATUS "_my_func")

    if (CALL_MYSELF)
        _my_func(OFF)
    endif()
endfunction()

function(my_func)
    message(STATUS "my_func")
    _my_func(ON)
endfunction()

my_func()
```
Tested now on three PCs with the latest CMake (2x Windows, 1x MacOs) - it worked. Really don't understand why exactly this is not working when I remove the added underscore. Once again the error message I get when executing the original code:
```
[cmake] CMake Error at cmake/omnet/GetNedFolders.cmake:31 (_get_ned_folders):
[cmake]   Maximum recursion depth of 1000 exceeded
[cmake] Call Stack (most recent call first):
[cmake]   cmake/omnet/GetNedFolders.cmake:31 (_get_ned_folders)
```

**Comment by HpLightcorner** [comment]
Of course, it's broken now XD
Let me find out what's missing...

EDIT: Nevermind, there were some test stubs in the CMake Code causing the error. It still works fine with 6.x and Windows :)

When we proceed with my suggestion for Artery, I will then also test 5.x on windows, but I think there won't be show stoppers.

File: GenerateOppMessage.cmake (hunk 5 above)
Comment (thor) [reply GenerateOppMessage.cmake]: I've seen `SRCS` too. On the other hand, `SOURCES` is also a [target property](https://cmake.org/cmake/help/latest/prop_tgt/SOURCES.html).

As long as we're consistent internally I'm all 👍, though on a non-CMake-related matter: I prefer fully spelled out variables `commands` over `cmds`, `table` over `tbl`, `directory` over `dir` (except in a maths related context, but that's different). Thus, in that sense I'd suggest as clear and straightforward variable names as possible, and prefer `SOURCES`.


(That being said, I'm sure I haven't always followed that preference myself, hehe.)

**Comment by thor** [comment]
I'm a bit preoccupied at the moment with a funeral, but 1) I'll test it relatively early next week, 2) feel free to merge this without waiting for my review, before I get around to testing my setup if you'd like, and then I'll 3) create issues for what comes around.

File: GenerateOppMessage.cmake (hunk 5 above)
Comment (riebl) [reply GenerateOppMessage.cmake]: The more I think about it, the more I am afraid that users may confuse "SRCS"/"SOURCES" with input sources.
But that might just be me getting paranoid…

File: GetNedFolders.cmake (hunk 8 above)
Comment (riebl) [reply GetNedFolders.cmake]: Renaming `_get_ned_folders` to `_get_ned_folders_plumbing` should work as well, does it? I would prefer this solution over double underscores.

File: GenerateOppMessage.cmake (hunk 5 above)
Comment (HpLightcorner) [reply GenerateOppMessage.cmake]: we could name the variables `GEN_SOURCES` and `GEN_INCLUDE_DIR`?

File: OMNet.cmake (hunk 7 above)
Comment (HpLightcorner) [reply OMNet.cmake]: It seems that CMake `file` module offers this feature for ages. I will change that syntax and set the version to 3.13.

**Comment by HpLightcorner** [comment]
@riebl - can you check if changing from `cmake_path` to `file` is working with your CMake version?

**Comment by riebl** [comment]
@HpLightcorner I have checked it with CMake 3.13.5 and it builds fine.

File: GenerateOppMessage.cmake (hunk 3 above)
Comment (HpLightcorner) [reply GenerateOppMessage.cmake:15]: @riebl - it is time to conclude - your final thoughts?

Comment (riebl) [reply GenerateOppMessage.cmake:15]: I am fine with leaving it as `ADDITIONAL_NED_PATHS` then.

**Comment by HpLightcorner** [comment]
@riebl and @thor - I guess we are ready to merge into develop and start with short-lived feature branches!

File: GenerateOppMessage.cmake [hunk 9]
```diff
@@ -1,31 +1,97 @@
...
+# - ADDITIONAL_NED_PATHS: Optional paths to be added as search/include directories when calling the message compiler (-I Arguments)
+
 # generate sources for messages via opp_msgc
 function(generate_opp_message msg_input)
-    cmake_parse_arguments(args "" "TARGET;DIRECTORY" "" ${ARGN})
+    set(options MSG4)
+    set(oneValueArgs TARGET DIRECTORY OUTPUT_ROOT GEN_SOURCES GEN_INCLUDE_DIR)
+    set(multiValueArgs ADDITIONAL_NED_PATHS)
+
+    cmake_parse_arguments(args "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
```
Comment (thor) [inline GenerateOppMessage.cmake]:
```suggestion
    set(options_args MSG4)
    set(single_args TARGET DIRECTORY OUTPUT_ROOT GEN_SOURCES GEN_INCLUDE_DIR)
    set(multi_args ADDITIONAL_NED_PATHS)

    cmake_parse_arguments(args "${options_args}" "${single_args}" "${multi_args}" ${ARGN})
```

**suggestion (non-blocking)**: use the same names for the argument variables as in `add_opp_target`. 

I wouldn't mind calling them something else entirely, although I'm not sure what I'd suggest. The bikeshedding in itself isn't too important. Consistency would be nice, however.
